File: algorithms/karatsuba/karatsuba.js
/**
 * Karatsuba algorithm for fast multiplication
 *
 * ab * cd == (a*c)^(m*2) + ((a+b)*(c+d) - a*c - b*d)^m + b*d
 * where m = max( ab.length, cd.length ) / 2
 * (a+b)(c+d) - ac - bd == ac + ad + bc + bd - ac - bd == ad + bc
 * @param { Number } x
 * @param { Number } y
 * @returns { Number }
 */
function karatsuba(x, y) {
  if (x < 10 || y < 10) return x * y;

  const n = Math.max(x.toString().length, y.toString().length);
  const m = Math.ceil(n / 2);

  const a = Math.floor(x / 10 ** m);
  const b = x % 10 ** m;
  const c = Math.floor(y / 10 ** m);
  const d = y % 10 ** m;

  const ac = karatsuba(a, c);
  const bd = karatsuba(b, d);
  const adbc = karatsuba(a + b, c + d) - ac - bd;

  return ac * 10 ** (m * 2) + adbc * 10 ** m + bd;
}

console.assert(karatsuba(2, 2) == 2 * 2, '2 * 2 == 4');
console.assert(karatsuba(0, 1) == 0 * 1, '0 * 1 == 0');
console.assert(karatsuba(1, 0) == 1 * 0, '1 * 0 == 0');
console.assert(karatsuba(1, 42) == 1 * 42, '1 * 42 == 42');
console.assert(karatsuba(1000, 64) == 1000 * 64, '1000 * 64 == 64000');
console.assert(karatsuba(10000, 1000) == 10000 * 1000, '10000 * 1000 == 10000000');
console.assert(karatsuba(1024, 2048) == 1024 * 2048, '1024 * 2048 == 2097152');

File: algorithms/karatsuba/karatsuba.py
from math import ceil, floor
def karatsuba(x, y):
  """Return x * y
  Karatsuba algorithm for fast multiplication

  >>> karatsuba(2, 2) == 2 * 2
  True
  >>> karatsuba(0, 1) == 0 * 1
  True
  >>> karatsuba(1, 0) == 1 * 0
  True
  >>> karatsuba(1, 42) == 1 * 42
  True
  >>> karatsuba(1000, 64) == 1000 * 64
  True
  >>> karatsuba(10000000, 1000000) == 10000000 * 1000000
  True
  >>> karatsuba(1024, 2048) == 1024 * 2048
  True
  """
  if x < 10 or y < 10: return x * y

  n = max(len(str(x)), len(str(y)))
  m = ceil(n / 2)
  
  a = floor(x / 10**m)
  b = x % 10**m
  c = floor(y / 10**m)
  d = y % 10**m

  ac = karatsuba(a, c)
  bd = karatsuba(b, d)
  adbc = karatsuba(a+b, c+d) - ac - bd

  return ac * 10**(m*2) + adbc * 10**m + bd

if __name__ == "__main__":
    import doctest
    doctest.testmod()